pennylane/gradients/param_shift_hessian.py

```diff
@@ -0,0 +1,320 @@
+    gradient_recipe = gradient_recipe[:, :, np.lexsort(np.abs(gradient_recipe)[:, -1])]
+
+    return gradient_recipe[0, 0], gradient_recipe[0, 1], gradient_recipe[:, 2].T
+
+
+def generate_multishifted_tapes(tape, idx, shifts):
+    r"""Generate a list of tapes where the corresponding trainable parameter
+    indices have been shifted by the values given.
+
+    Args:
+        tape (.QuantumTape): input quantum tape
+        idx (list[int]): trainable parameter indices to shift the parameters of
+        shifts (list[list[float or int]]): nested list of shift values, each
+            list containing a value for each index
+
+    Returns:
+        list[QuantumTape]: List of quantum tapes. Each tape has multiple parameters
+            (indicated by ``idx``) shifted by the values of ``shifts``. The length
+            of the returned list of tapes will match the length of ``shifts``.
+    """
+    params = list(tape.get_parameters())
+    tapes = []
+
+    for shift in shifts:
+        new_params = params.copy()
+        shifted_tape = tape.copy(copy_operations=True)
+
+        for i, s in enumerate(shift):
+            new_params[idx[i]] = new_params[idx[i]] + qml.math.convert_like(s, new_params[idx[i]])
+
+        shifted_tape.set_parameters(new_params)
+        tapes.append(shifted_tape)
+
+    return tapes
+
+
+class hessian_transform(qml.batch_transform):
+    """Modified `gradient_transform` to account for different post-processing when computing
+    the second derivate of a QNode."""
+
+    def __init__(
+        self, transform_fn, expand_fn=expand_invalid_trainable, differentiable=True, hybrid=True
+    ):
+        self.hybrid = hybrid
+        super().__init__(transform_fn, expand_fn=expand_fn, differentiable=differentiable)
+
+    def default_qnode_wrapper(self, qnode, targs, tkwargs):
+        # Here, we overwrite the QNode execution wrapper in order
+        # to take into account that classical processing may be present
+        # inside the QNode.
+        hybrid = tkwargs.pop("hybrid", self.hybrid)
+        _wrapper = super().default_qnode_wrapper(qnode, targs, tkwargs)
+        cjac_fn = qml.transforms.classical_jacobian(qnode, expand_fn=expand_invalid_trainable)
+
+        def jacobian_wrapper(*args, **kwargs):
+            qjac = _wrapper(*args, **kwargs)
+
+            if any(m.return_type is qml.operation.Probability for m in qnode.qtape.measurements):
+                qjac = qml.math.squeeze(qjac)
+
+            if not hybrid:
+                return qjac
+
+            kwargs.pop("shots", False)
+            cjac = cjac_fn(*args, **kwargs)
+
+            if isinstance(cjac, tuple):
+                # Classical processing of multiple arguments is present. Return qjac @ cjac.
+                jacs = [
+                    qml.math.squeeze(qml.math.tensordot(c, qjac, [[0], [-1]]))
+                    for c in cjac
+                    if c is not None
+                ]
```

Comment:
I think this is causing the following program to fail:

```python
import pennylane as qml
import tensorflow as tf

dev = qml.device("default.qubit", wires=3)

@qml.qnode(dev, interface="tf", diff_method="parameter-shift", max_diff=2)
def circuit(weights):
    qml.RX(weights[0], wires=0)
    qml.RY(weights[1], wires=0)
    qml.CNOT(wires=[0, 1])
    qml.RZ(weights[2], wires=1)
    return qml.probs(wires=0)


weights = tf.Variable([0.1, 0.2, 0.3], dtype=tf.float64)

with tf.GradientTape() as tape:
    loss = circuit(weights)

print(loss)

hess = qml.gradients.param_shift_hessian(circuit)(weights)
print(hess)
```

Reply:
with
```
Traceback (most recent call last):
  File "test6.py", line 23, in <module>
    hess = qml.gradients.param_shift_hessian(circuit)(weights)
  File "/home/josh/xanadu/pennylane/pennylane/gradients/param_shift_hessian.py", line 103, in jacobian_wrapper
    jacs = [
  File "/home/josh/xanadu/pennylane/pennylane/gradients/param_shift_hessian.py", line 104, in <listcomp>
    qml.math.squeeze(qml.math.tensordot(c, qjac, [[0], [-1]]))
  File "/home/josh/miniconda3/lib/python3.8/site-packages/autoray/autoray.py", line 84, in do
    return get_lib_fn(backend, fn)(*args, **kwargs)
  File "/home/josh/miniconda3/lib/python3.8/site-packages/tensorflow/python/util/traceback_utils.py", line 153, in error_handler
    raise e.with_traceback(filtered_tb) from None
  File "/home/josh/miniconda3/lib/python3.8/site-packages/tensorflow/python/framework/ops.py", line 7107, in raise_from_not_ok_status
    raise core._status_to_exception(e) from None  # pylint: disable=protected-access
tensorflow.python.framework.errors_impl.InvalidArgumentError: Matrix size-incompatible: In[0]: [3,3], In[1]: [0,1] [Op:MatMul]
```

Reply:
Good catch @josh146! I forgot to come back to these lines and they aren't adequately covered by the testing apparently.

To be honest, I was wondering why we differentiate this case, it's not clear to me how is this execution path is triggered. It's not the case with your example above, but I think I determined it's for the case where there are multiple QNode arguments of different size, where a single Jacobian matrix can't be used?
In which case it would be beneficial to add tests that cover this scenario.

Reply:
Just find this page which explains it https://pennylane.readthedocs.io/en/stable/code/api/pennylane.transforms.classical_jacobian.html

Reply:
Yep, that's the one! Surprised my example above wasn't triggering this branch 🤔

Reply:
Actually, I realized that this error is not specific to the hessian transform, as the same happens with the gradient transform. Apparently with the TensorFlow interface, parameters won't be identified as trainable unless the differentiation happens within a gradient tape:
```py
import pennylane as qml
import tensorflow as tf

dev = qml.device("default.qubit", wires=3)

@qml.qnode(dev, interface="tf", diff_method="parameter-shift", max_diff=2)
def circuit(weights):
    qml.RX(weights[0], wires=0)
    qml.RY(weights[1], wires=0)
    qml.CNOT(wires=[0, 1])
    qml.RZ(weights[2], wires=1)
    return qml.probs(wires=0)

weights = tf.Variable([0.1, 0.2, 0.3], dtype=tf.float64)

jac = qml.gradients.param_shift(circuit)(weights)
```
fails, while
```py
with tf.GradientTape() as tape:
    jac = qml.gradients.param_shift(circuit)(weights)
```
does not.

With no trainable parameter the quantum Jacobian is an empty matrix, leading to the error message.

Reply:
Oh!! Yes, this is a known issue (and affects the metric tensor as well! see https://github.com/PennyLaneAI/pennylane/issues/2040#issuecomment-996736526).

Since this is an issue already in the `master` branch, no need to address it in this PR.